### PR TITLE
Prototype

### PR DESCRIPTION
**Pull request by andrei-marinica**
Initial prototype.

Has support for wasmer imports. Can run the factorial smart contract.

### CODE REVIEW

File: exec-c-api/build.rs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+"#
+    .to_string();
+
+    // Generate the C bindings in the `OUT_DIR`.
+    out_wasmer_header_file.set_extension("h");
+    Builder::new()
+        .with_crate(crate_dir.clone())
+        .with_language(Language::C)
+        .with_header(&pre_header)
+        .with_define("target_family", "windows", "_WIN32")
```
[the remark this answers is not in the document]
Comment (andrei-marinica) [reply exec-c-api/build.rs:56]: It's copied from wasmer 1. We can discuss, it is an interesting topic

File: exec-c-api/src/capi_instance.rs [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+    wasm_bytes_ptr: *mut u8,
+    wasm_bytes_len: u32,
+    options_ptr: *const vm_exec_compilation_options_t,
+) -> vm_exec_result_t {
+    // unpack the executor object
+    if executor.is_null() {
+        with_service(|service| service.update_last_error_str("executor ptr is null".to_string()));
+        return vm_exec_result_t::VM_EXEC_ERROR;
+    }
+    let capi_executor = &mut *(executor as *mut CapiExecutor);
```
[the remark this answers is not in the document]
Comment (andrei-marinica) [reply exec-c-api/src/capi_instance.rs:69]: The return makes it tricky to make into a function, but I can do a macro, I guess

File: exec-c-api/src/capi_instance.rs [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+/// // the standard output.
+/// void print(wasmer_instance_context_t *context, int32_t pointer, int32_t length) {
+///     // Use `wasmer_instance_context` to get back the first instance memory.
+///     const wasmer_memory_t *memory = wasmer_instance_context_memory(context, 0);
+///
+///     // Continue…
+/// }
+/// ```
+// #[repr(C)]
+// pub struct wasmer_instance_context_t;
```
[the remark this answers is not in the document]
Comment (andrei-marinica) [reply exec-c-api/src/capi_instance.rs:46]: I think I am going to need this object when dealing with memory. I am keeping the commented code until then.

File: exec-c-api/src/capi_instance.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (andrei-marinica) [reply exec-c-api/src/capi_instance.rs]: I did it with some macros.

File: exec-c-api/src/capi_instance.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (andrei-marinica) [reply exec-c-api/src/capi_instance.rs:46]: I am keeping this for the next PR, it seems like I might need the `wasmer_instance_context_t` for memory handling

File: exec-c-api/src/capi_executor.rs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+    // create executor
+    let executor_result =
+        with_service(|service| service.new_executor(Box::new(CapiVMHooks::new(vm_hook_pointers))));
+    match executor_result {
+        Ok(executor_box) => {
+            let capi_executor = CapiExecutor {
+                content: executor_box,
+            };
+            *executor = Box::into_raw(Box::new(capi_executor)) as *mut vm_exec_executor_t;
```
[the remark this answers is not in the document]
Comment (andrei-marinica) [reply exec-c-api/src/capi_executor.rs:37]: `Box::into_raw` leaks memory, the new owner needs to deallocate manually.

**Review by ovstinga — APPROVED** [review]

**Review by camilbancioiu — APPROVED** [review]

**Pull request merged**